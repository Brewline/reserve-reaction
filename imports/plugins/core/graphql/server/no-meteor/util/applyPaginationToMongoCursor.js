const DEFAULT_LIMIT = 20;

/**
 * Inspired by https://www.reindex.io/blog/relay-graphql-pagination-with-mongodb/
 * @name applyPaginationToMongoCursor
 * @method
 * @memberof GraphQL/ResolverUtilities
 * @summary Adds `skip` and `limit` to a MongoDB cursor as necessary, based on GraphQL
 *   `first` and `last` params
 * @param {Cursor} cursor MongoDB cursor
 * @param {Object} args GraphQL query arguments
 * @param {Number} totalCount Total count of docs that match the query, after applying the before/after filter
 * @return {Promise<Object>} `{ pageInfo }`
 */
export default async function applyPaginationToMongoCursor(cursor, { first, last } = {}, totalCount) {
  if (first && last) throw new Error("Request either `first` or `last` but not both");

  // Enforce a `first: 20` limit if no user-supplied limit, using the DEFAULT_LIMIT
  const limit = first || last || DEFAULT_LIMIT;

  let skip = 0;
  if (last && totalCount > last) skip = totalCount - last;

  let hasNextPage = null;
  let hasPreviousPage = null;
  if (last) {
    if (skip === 0) {
      hasPreviousPage = false;
    } else {
      // For backward pagination, we can find out whether there is a previous page here, but we can't
      // find out whether there's a next page because the cursor has already had "before" filtering
      // added. Code external to this function will need to determine whether there are any documents
      // after that "before" ID.
      const prevCursor = cursor.clone();
      prevCursor.limit(limit + 1);
      prevCursor.skip(skip - 1);
      const prevCursorCount = await prevCursor.count();
      hasPreviousPage = prevCursorCount > limit;
    }
  } else {
    // For forward pagination, we can find out whether there is a next page here, but we can't
    // find out whether there's a previous page because the cursor has already had "after" filtering
    // added. Code external to this function will need to determine whether there are any documents
    // before that "after" ID.
    const nextCursor = cursor.clone();
    nextCursor.limit(limit + 1);
    const nextCursorCount = await nextCursor.count();
    hasNextPage = nextCursorCount > limit;
  }

  // Now apply actual limit + skip to the provided cursor
  cursor.limit(limit);
  if (skip) cursor.skip(skip);

  return {
    hasNextPage,
    hasPreviousPage
  };
}
