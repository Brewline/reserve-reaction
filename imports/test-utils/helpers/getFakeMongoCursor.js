/**
 * Helper functions for use in Jest tests
 * @namespace TestHelpers
 */

/**
 * @name getFakeMongoCursor
 * @method
 * @memberof TestHelpers
 * @param {String} collectionName - name of the collection
 * @param {Any} results - results to be returned as part of the cursor
 * @return {Object} fake cursor
 */
export default function getFakeMongoCursor(collectionName, results, options) {
  const cursor = {
    cmd: {
      query: (options && options.query) || {}
    },
    ns: `meteor.${collectionName}`,
    options: {
      db: {
        databaseName: "meteor"
      }
    }
  };

  const isTesting = typeof jest !== "undefined";

  if (isTesting) {
    cursor.clone = jest.fn().mockName("cursor.clone").mockImplementation(() => cursor);
    cursor.count = jest.fn().mockName("cursor.count").mockReturnValue(results.length);
    cursor.filter = jest.fn().mockName("cursor.filter").mockReturnValue(cursor);
    cursor.limit = jest.fn().mockName("cursor.limit").mockReturnValue(cursor);
    cursor.options.db.collection = jest.fn().mockName("cursor.options.db.collection").mockReturnValue({
      findOne: jest.fn().mockName("cursor.options.db.collection.findOne").mockResolvedValue(null)
    });
    cursor.skip = jest.fn().mockName("cursor.skip").mockReturnValue(cursor);
    cursor.sort = jest.fn().mockName("cursor.sort").mockReturnValue(cursor);
    cursor.toArray = jest.fn().mockName("cursor.toArray").mockResolvedValue(results);
  } else {
    cursor.clone = () => cursor;
    cursor.count = () => results.length;
    cursor.filter = () => cursor;
    cursor.limit = () => cursor;
    cursor.options.db.collection = () => ({
      findOne: () => Promise.resolve(null)
    });
    cursor.skip = () => cursor;
    cursor.sort = () => cursor;
    cursor.toArray = () => Promise.resolve(results);
  }

  return cursor;
}
