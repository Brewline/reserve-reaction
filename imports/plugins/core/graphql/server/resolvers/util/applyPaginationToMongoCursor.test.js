import applyPaginationToMongoCursor from "./applyPaginationToMongoCursor";
import getFakeMongoCursor from "/imports/test-utils/helpers/getFakeMongoCursor";

let mockCursor;
beforeEach(() => {
  mockCursor = getFakeMongoCursor("Test", new Array(100));
});

test("with neither first nor last limits to first 20", async () => {
  mockCursor.count.mockReturnValueOnce(Promise.resolve(21));
  const result = await applyPaginationToMongoCursor(mockCursor, undefined, 100);
  expect(result).toEqual({
    hasNextPage: true,
    hasPreviousPage: null
  });
  expect(mockCursor.limit.mock.calls).toEqual([[21], [20]]);
  expect(mockCursor.skip).not.toHaveBeenCalled();
});

test("with both first and last, throws error", () => {
  expect(applyPaginationToMongoCursor(mockCursor, { first: 1, last: 1 }, 100)).rejects.toThrowErrorMatchingSnapshot();
});

test("with first and more, returns hasNextPage true", async () => {
  mockCursor.count.mockReturnValueOnce(Promise.resolve(51));
  const result = await applyPaginationToMongoCursor(mockCursor, { first: 50 }, 100);
  expect(result).toEqual({
    hasNextPage: true,
    hasPreviousPage: null
  });
  expect(mockCursor.limit.mock.calls).toEqual([[51], [50]]);
  expect(mockCursor.skip).not.toHaveBeenCalled();
});

test("with first and no more, returns hasNextPage false", async () => {
  mockCursor.count.mockReturnValueOnce(Promise.resolve(50));
  const result = await applyPaginationToMongoCursor(mockCursor, { first: 50 }, 100);
  expect(result).toEqual({
    hasNextPage: false,
    hasPreviousPage: null
  });
  expect(mockCursor.limit.mock.calls).toEqual([[51], [50]]);
  expect(mockCursor.skip).not.toHaveBeenCalled();
});

test("with last and more, returns hasPreviousPage true", async () => {
  mockCursor.count.mockReturnValueOnce(Promise.resolve(51));
  const result = await applyPaginationToMongoCursor(mockCursor, { last: 50 }, 80);
  expect(result).toEqual({
    hasNextPage: null,
    hasPreviousPage: true
  });
  expect(mockCursor.limit.mock.calls).toEqual([[51], [50]]);
  expect(mockCursor.skip.mock.calls).toEqual([[29], [30]]);
});

test("with last and no more, returns hasPreviousPage false", async () => {
  mockCursor.count.mockReturnValueOnce(Promise.resolve(50));
  const result = await applyPaginationToMongoCursor(mockCursor, { last: 50 }, 80);
  expect(result).toEqual({
    hasNextPage: null,
    hasPreviousPage: false
  });
  expect(mockCursor.limit.mock.calls).toEqual([[51], [50]]);
  expect(mockCursor.skip.mock.calls).toEqual([[29], [30]]);
});
