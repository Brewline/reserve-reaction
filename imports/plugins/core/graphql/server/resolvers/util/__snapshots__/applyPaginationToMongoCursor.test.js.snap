// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with both first and last, throws error 1`] = `"Request either \`first\` or \`last\` but not both"`;
